### ISSUE TITLE
Tests do not capture stdout()

### ISSUE THREAD

**Issue by Khouderchah-Alex** [rust-lang/rust#90785]
Output capture for tests (https://github.com/rust-lang/rust/commit/d0d0e7820806bbb89a2733bae8f9aaa9c78345d0, https://github.com/rust-lang/rust/commit/11ce918c75b05d065ce3bf98b20d62465b5afc55) causes print_to (and consequently _print & macros) to redirect to OUTPUT_CAPTURE if set. IIUC, this leaves any direct usage of stdout() uncaptured.

This is inconvenient for systems which use stdout() with generic code over io::Write.

Basically a dupe of the unresolved-but-closed #12309.

**Comment by mina86** [comment]
Steps to reproduce:

```console
$ cat src/lib.rs
#[test]
fn test() {
    use std::io::Write;

    eprintln!("eprintln");
    println!("println");
    std::io::stdout().write_all(b"stdout\n").unwrap();
    std::io::stderr().write_all(b"stderr\n").unwrap();
}
$ cargo test --tests
    Finished test [unoptimized] target(s) in 0.00s
     Running unittests (target/debug/deps/a-7d5a981fc2a1d521)

running 1 test
stdout
stderr
test test ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

$ cargo --version
cargo 1.58.0-nightly (7f08ace4f 2021-11-24)
```

Expected output:

```console
$ cargo test --tests
    Finished test [unoptimized] target(s) in 0.00s
     Running unittests (target/debug/deps/a-7d5a981fc2a1d521)

running 1 test
test test ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

Without looking at libtest code, the proper solution seems to be:

```c
old_stdout = dup(1);
old_stderr = dup(2);

tmp = mktemp(…);
tmp_fd = open(tmp);

dup2(tmp_fd, 1);
dup2(tmp_fd, 2);

success = run_the_test();

dup2(old_stdout, 1);
dup2(old_stderr, 2);

if !success {
    fseek(tmp_fd, 0);
    cat(tmp_fd);
}

close(tmp_fd);
unlink(tmp);
```

This would also solve https://github.com/rust-lang/rust/issues/92370

**Comment by davepacheco** [comment]
I ran into this today as well.  I'm not sure how hard this is to fix.  Even if it's decided not to fix it, I think some better documentation around output capturing could help.  I understood it to be stdout/stderr that was captured, but it's really just thread-local `print!` and friends.  I ran into this because I was using `term::stdout()`, but I think you'd also be likely to see this in async tests that print output from other tasks.

**Comment by vojtechkral** [comment]
Ran into this today. This is particularly unpleasant in the following scenario (pseudo-Rust):

```
eprintln!("Something something");
let _ = stderr.write_all(tty_escape_cursor_up_clear_line);
eprintln!("Something else");
```

running that code from the test harness leads to the eprintlns being capture, but the escape sequences not so much, and they then end up messing up the test harness output.

The worst bit is I can't seem to find a way to detect this somehow (in test the stderr is still a tty), I'm probably going to have to introduce some kind of switch to disable tty stuff from tests explicitly.

**Comment by mathstuf** [comment]
> Without looking at libtest code, the proper solution seems to be: [[`dup` juggling]]

Alas, `dup` and friends are process-wide, so this strategy forces a limit of `-j1` on test execution.

**Comment by vojtechkral** [comment]
> > Without looking at libtest code, the proper solution seems to be: [[`dup` juggling]]
> 
> Alas, `dup` and friends are process-wide, so this strategy forces a limit of `-j1` on test execution.

Is that a problem though? Arguably you could dup original 1 and 2, use that for all logging/cli output and make 1 and 2 pipes.

**Comment by vojtechkral** [comment]
Anyone knows what nextest does? Chances are they've got this solved already...

**Comment by mathstuf** [comment]
> make 1 and 2 pipes (for tests to write to).

Pipes to what? There still wouldn't be per-test places for things to go, so "what happened for failed tests" would be jumbled.

**Comment by vojtechkral** [comment]
> Pipes to what?
A thread (not) copying that to the real std{out,err} based on settings. It could also filter out escape sequences.

> There still wouldn't be per-test places for things to go, so "what happened for failed tests" would be jumbled.
Sure, but at least the output would actually be captured. And you can always use `-j1` when tests fail and you need clean output.

**Comment by AlejandroFNadal** [comment]
For people searching and finding this issue:

Add in each test, instead of
`#[test]`
put:

`#[test_log::test]`

Then, to cargo toml
```
[dev-dependencies]
test-log = "*"
```

Finally, set env variable to desired debugging level:
`export RUST_LOG="debug"`
